src/openml.jl

```diff
@@ -47,41 +48,30 @@ Receives an `HTTP.Message.response` that has an
 ARFF file format in the `body` of the `Message`.
 """
 function convert_ARFF_to_rowtable(response)
-    data = String(response.body)
-    data2 = split(data, "\n")
-
     featureNames = String[]
     dataTypes = String[]
-    # TODO: make this more performant by anticipating types?
-    named_tuples = [] # `Any` type here bad
-    for line in data2
+    io = IOBuffer(response.body)
+    for line in eachline(io)
         if length(line) > 0
             if line[1:1] != "%"
                 d = []
                 if occursin("@attribute", lowercase(line))
-                    push!(featureNames, replace(replace(split(line, " ")[2], "'" => ""), "-" => "_"))
-                    push!(dataTypes, split(line, " ")[3])
+                    splitline = split(line)
+                    push!(featureNames, replace(replace(splitline[2], "'" => ""), "-" => "_"))
+                    push!(dataTypes, splitline[3])
                 elseif occursin("@relation", lowercase(line))
                     nothing
                 elseif occursin("@data", lowercase(line))
                     # it means the data starts
-                    nothing
-                else
-                    values = split(line, ",")
-                    for i in eachindex(featureNames)
-                        if lowercase(dataTypes[i]) in ["real","numeric"]
```

Comment:
So, as I understand it, the old code here makes use of metadata `dataTypes` about types ("real", "numeric", ...) to inform the data parsing. Are you simply ignoring this information here and attempting to use the CSV parser to infer all types from the raw data itself? I would have thought we should rely on this metadata to the extent that we can, no? (But maybe this metadata is actually not that reliable anyway, if I remember correctly?)

Reply:
Yes, the code currently ignores this metadata; the CSV parser infers the type.
One option would be to check the consistency of the inferred and metadata types and issue a warning upon mismatch.
I did this in a little test which warned me mostly about multiclass features being occasionally inferred as Int, e.g.
```julia
julia> data = MLJOpenML.load(8);
┌ Warning: Inferred Int64 instead of {1,2} for feature selector.
│          Please coerce to the desired type manually.
└ @ MLJOpenML ~/MLJOpenML.jl/src/openml.jl:46
```
If we add CategoricalArrays as a dependency I guess we could fix this. Do you want this, or shall we just warn the user?

Reply:
Okay, I looked a bit more into this (catching up with you, really). Notes to self:

Each feature has a type in the **metadata** (what is called `dataTypes[i]` on dev branch), a corresponding type **displayed** on the OpenML website GUI, and a  **CSV** inferred type. 

Unfortunately, I don't believe we can reliably infer "true" scientific types from the metadata type in all cases because: 

1.  The metadata type does not distinguish between `OrderedFactor` and `Count` (it seems INTEGER is used for both), and
2. `STRING` is sometimes used as a designation for data that is probably `Multiclass`, sometimes `OrderedFactor`, and sometimes `Textual`, but I'm not 100% sure about this. 

metdata  | displayed | scientific type
---------|------------|--------------
`INTEGER` or `Integer` or `integer` | `numeric` | `OrderedFactor` (sometimes `Count`)
`REAL` or `Real` or `real` | `numeric` | `Continuous`
`STRING` or `String` or `string` | `string` | ?
`{class1, ...., classk}` | `nominal` | `Multiclass` (sometimes `OrderedFactor`)

At present there is nothing particularly "MLJ" about "MLJOpenML" and we should not force users into the `ScientificTypes` mould, but it should be there as an option, I reckon.

---

So, my suggestion is that we add a keyword `parser=...` with options: `:csv` (fastest), `:openml` (default),  and `:auto_type` (for best scientific type interpretation):

- `:csv` - use the CSV parser and perform the following checks:
    - `INTEGER` types are `<:Integer`, `REAL` types are `<:AbstractFloat`, `STRING` types are   `AbstractString` and there exist no `{...}` types. If any check fails, this is reported to `@info`; something like "The following features do not have types matching those in the OpenML metadata: ... To force type coercion, including the use of unordered `CategoricalArrays` for nominal data, specify `parser=:openml`. To suppress this message, specify `verbosity=0`. " (and add `verbosity` keyword arg).
    
   In this mode however, no type coercion is performed and a `CSV.File` object is returned. 

- `:openml` -  The same as the above, except the necessary type coercions are performed and nothing is reported. One simple way to to implement the type coercion, would to use `ScientificTypes.coerce`, as in  `new_table = coerce(csv_file_table, :feature1 => Count, :=> :feature2 => Continuous, :feature3 => Multiclass, ...)` where `feature1` is wanting to be `INTEGER`, `feature2` is wanting to be `REAL` and `feature3` is wanting to be `{...}`. (This will automatically handle missing values appropriately.) At present, this involves making an in-memory copy of all the data, ie coercion is not in-place. An alternative that could be use later is to materialized the file as column dictionary and use `coerce!` after  https://github.com/JuliaAI/ScientificTypes.jl/issues/136 is resolved. 

- `auto` - Return `ScientificTypes.autotype(table)` where `table` is the output of the `:openml` option. 

@jbrea What do you think?

Reply:
That's a great write-up, thanks!

I implemented the :csv and :openml parser. Currently, the :openml parser can be very slow, due to slow coercion, e.g. for data set 1088, but it works reasonably well even for MNIST (data set 554).

I didn't understand what `parser = :auto` should do. I tried for example `ScientificTypes.autotype(MLJOpenML.load(554, parser = :openml))` but I couldn't make much sense of the resulting dictionary.

Reply:
Thanks for extra mile you have gone here with the new commits. I'll have a final look after the weekend.

>  but I couldn't make much sense of the resulting dictionary.

Oops, my bad. You need a second step that applies the coercions suggested by the dictionary. So, return `coerce(table, autotype(table))`, where `table` is output of `:openml` option. The function `autotype` tries to infer type coercions on the basis of some heuristics documented [here](https://juliaai.github.io/ScientificTypes.jl/dev/#Automatic-type-conversion-for-tabular-data-1). It's not terribly sophisticated but could be improved in the future, introducing some ML (must be project like this somewhere). In any case I would *not* make `:auto` the default. In some sense `:csv` would be a good option because it's fast and if it's "wrong" it reports this to the user who can take further action if she really wants to, while the other option might frustrate users in some cases, without them realising they have a faster option. I don't have a strong opinion however.

Reply:
One other suggestion is that we pass any extra kwargs as kwargs to the csv parser. Thanks.

Reply:
Great, thanks. I ran quite a few more tests with randomly sampled datasets, fixed some small issues and added support for sparse ARFF files. This should now be ready to be merged.